File: 74751_lab5_task5.py
import copy

class PatientMonitor:
    def __init__(self, monitor_id="UNKNOWN", patient_id="UNKNOWN", obj=None):
        if obj is not None and isinstance(obj, PatientMonitor):
            self.monitor_id = obj.monitor_id + "_copy"
            self.patient_id = obj.patient_id
            self.vital_signals = copy.deepcopy(obj.vital_signals)
        else:
            self.monitor_id = monitor_id
            self.patient_id = patient_id
            self.vital_signals = {}

    def __del__(self):
        print(f"Patient Monitor {self.monitor_id} detached from patient {self.patient_id}.")

    def add_signal_sample(self, channel, value):
        if channel not in self.vital_signals:
            self.vital_signals[channel] = []
        self.vital_signals[channel].append(value)

    def compute_average(self, channel):
        if channel in self.vital_signals and self.vital_signals[channel]:
            return sum(self.vital_signals[channel]) / len(self.vital_signals[channel])
        return 0.0

    def detect_abnormal_readings(self, channel, min_safe, max_safe):
        if channel in self.vital_signals:
            for val in self.vital_signals[channel]:
                if val < min_safe or val > max_safe:
                    print(f"ALERT: Abnormal reading {val} detected on channel {channel}!")

    def produce_monitoring_summary_report(self):
        print("--- Monitoring Summary Report ---")
        print(f"Monitor ID: {self.monitor_id} | Patient ID: {self.patient_id}")
        for channel, samples in self.vital_signals.items():
            print(f"Channel: {channel}")
            print(f"  - Samples: {len(samples)}")
            avg = sum(samples) / len(samples) if samples else 0.0
            print(f"  - Average: {avg:.2f}")
        print("---------------------------------")

if __name__ == "__main__":
    p1 = PatientMonitor("PM-01", "PT-1234")
    p1.add_signal_sample("HeartRate", 75.0)
    p1.add_signal_sample("HeartRate", 78.0)
    p1.add_signal_sample("HeartRate", 120.0) # Abnormal
    
    p1.add_signal_sample("Voltage", 12.1)
    p1.add_signal_sample("Voltage", 11.9)
    
    p2 = PatientMonitor(obj=p1) # Deep copy
    
    p1.detect_abnormal_readings("HeartRate", 60.0, 100.0)
    p1.produce_monitoring_summary_report()

File: 74751_lab5_task5.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <numeric>

using namespace std;

class PatientMonitor {
private:
    string monitorID;
    string patientID;
    map<string, vector<double>> vitalSignals;

public:
    // Default constructor
    PatientMonitor() : monitorID("UNKNOWN"), patientID("UNKNOWN") {}

    // Parameterized constructor
    PatientMonitor(string mID, string pID) : monitorID(mID), patientID(pID) {}

    // Deep-copy constructor
    PatientMonitor(const PatientMonitor& other) 
        : monitorID(other.monitorID + "_copy"), patientID(other.patientID) {
        for (const auto& pair : other.vitalSignals) {
            vitalSignals[pair.first] = vector<double>(pair.second);
        }
    }

    // Destructor
    ~PatientMonitor() {
        cout << "Patient Monitor " << monitorID << " detached from patient " << patientID << "." << endl;
    }

    void addSignalSample(string channel, double value) {
        vitalSignals[channel].push_back(value);
    }

    double computeAverage(string channel) const {
        auto it = vitalSignals.find(channel);
        if (it != vitalSignals.end() && !it->second.empty()) {
            double sum = accumulate(it->second.begin(), it->second.end(), 0.0);
            return sum / it->second.size();
        }
        return 0.0;
    }

    void detectAbnormalReadings(string channel, double minSafe, double maxSafe) const {
        auto it = vitalSignals.find(channel);
        if (it != vitalSignals.end()) {
            for (double val : it->second) {
                if (val < minSafe || val > maxSafe) {
                    cout << "ALERT: Abnormal reading " << val << " detected on channel " << channel << "!" << endl;
                }
            }
        }
    }

    void produceMonitoringSummaryReport() const {
        cout << "--- Monitoring Summary Report ---" << endl;
        cout << "Monitor ID: " << monitorID << " | Patient ID: " << patientID << endl;
        for (const auto& pair : vitalSignals) {
            cout << "Channel: " << pair.first << endl;
            cout << "  - Samples: " << pair.second.size() << endl;
            double sum = accumulate(pair.second.begin(), pair.second.end(), 0.0);
            double avg = pair.second.empty() ? 0.0 : sum / pair.second.size();
            cout << "  - Average: " << avg << endl;
        }
        cout << "---------------------------------" << endl;
    }
};

int main() {
    PatientMonitor p1("PM-01", "PT-1234");
    p1.addSignalSample("HeartRate", 75.0);
    p1.addSignalSample("HeartRate", 78.0);
    p1.addSignalSample("HeartRate", 120.0); // Abnormal
    
    p1.addSignalSample("Voltage", 12.1);
    p1.addSignalSample("Voltage", 11.9);

    PatientMonitor p2 = p1; // Deep copy
    
    p1.detectAbnormalReadings("HeartRate", 60.0, 100.0);
    p1.produceMonitoringSummaryReport();

    return 0;
}
